### PR TITLE
python311Packages.python-lsp-black: fix tests for black 24.3.0+

### LINKED ISSUE #302902: python311Packages.python-lsp-black: build failure, failing tests
### Describe the bug

As of master @ caacb878cdc69d7f6addfe8b526d0b664261b576, this package is failing to build.

This seems related to https://github.com/NixOS/nixpkgs/commit/f8e8b40db295aae1215cc8a38f6cae7502bdfb47

### Steps To Reproduce
1. `nix-build -A python311Packages.python-lsp-black`

Result:

```

Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-8.0.2, pluggy-1.4.0
rootdir: /build/source
collected 21 items                                                             

tests/test_plugin.py ....F......F.........                               [100%]

=================================== FAILURES ===================================
___________________ test_pylsp_format_document_syntax_error ____________________

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
>           src_ast = parse_ast(src)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1526: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n'

    def parse_ast(src: str) -> ast.AST:
        # TODO: support Python 4+ ;)
        versions = [(3, minor) for minor in range(3, sys.version_info[1] + 1)]
    
        first_error = ""
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=True)
            except SyntaxError as e:
                if not first_error:
                    first_error = str(e)
    
        # Try to parse without type comments
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=False)
            except SyntaxError:
                pass
    
>       raise SyntaxError(first_error)
E       SyntaxError: unexpected indent (<unknown>, line 1)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/parsing.py:148: SyntaxError

The above exception was the direct cause of the following exception:

config = <pylsp.config.config.Config object at 0x7ffff5632990>
invalid_document = <pylsp.workspace.Document object at 0x7ffff563fc10>

    def test_pylsp_format_document_syntax_error(config, invalid_document):
>       result = pylsp_format_document(config, invalid_document)

tests/test_plugin.py:179: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylsp_black/plugin.py:38: in pylsp_format_document
    return format_document(config, document)
pylsp_black/plugin.py:73: in format_document
    formatted_text = format_text(text=text, config=config, lines=lines)
pylsp_black/plugin.py:114: in format_text
    formatted_text = black.format_file_contents(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1083: in format_file_contents
    check_stability_and_equivalence(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1057: in check_stability_and_equivalence
    assert_equivalent(src_contents, dst_contents)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
            src_ast = parse_ast(src)
        except Exception as exc:
>           raise ASTSafetyError(
                "cannot use --safe with this file; failed to parse source file AST: "
                f"{exc}\n"
                "This could be caused by running Black with an older Python version "
                "that does not support new syntax used in your source file."
            ) from exc
E           black.parsing.ASTSafetyError: cannot use --safe with this file; failed to parse source file AST: unexpected indent (<unknown>, line 1)
E           This could be caused by running Black with an older Python version that does not support new syntax used in your source file.

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1528: ASTSafetyError
_____________________ test_pylsp_format_range_syntax_error _____________________

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
>           src_ast = parse_ast(src)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1526: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n'

    def parse_ast(src: str) -> ast.AST:
        # TODO: support Python 4+ ;)
        versions = [(3, minor) for minor in range(3, sys.version_info[1] + 1)]
    
        first_error = ""
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=True)
            except SyntaxError as e:
                if not first_error:
                    first_error = str(e)
    
        # Try to parse without type comments
        for version in sorted(versions, reverse=True):
            try:
                return _parse_single_version(src, version, type_comments=False)
            except SyntaxError:
                pass
    
>       raise SyntaxError(first_error)
E       SyntaxError: unexpected indent (<unknown>, line 1)

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/parsing.py:148: SyntaxError

The above exception was the direct cause of the following exception:

config = <pylsp.config.config.Config object at 0x7ffff5634950>
invalid_document = <pylsp.workspace.Document object at 0x7ffff567f610>

    def test_pylsp_format_range_syntax_error(config, invalid_document):
        range = {"start": {"line": 0, "character": 0}, "end": {"line": 1, "character": 0}}
    
>       result = pylsp_format_range(config, invalid_document, range=range)

tests/test_plugin.py:253: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylsp_black/plugin.py:46: in pylsp_format_range
    return format_document(config, document, range)
pylsp_black/plugin.py:73: in format_document
    formatted_text = format_text(text=text, config=config, lines=lines)
pylsp_black/plugin.py:114: in format_text
    formatted_text = black.format_file_contents(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1083: in format_file_contents
    check_stability_and_equivalence(
/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1057: in check_stability_and_equivalence
    assert_equivalent(src_contents, dst_contents)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src = '    x = 1+2\n', dst = 'x = 1 + 2\n'

    def assert_equivalent(src: str, dst: str) -> None:
        """Raise AssertionError if `src` and `dst` aren't equivalent."""
        try:
            src_ast = parse_ast(src)
        except Exception as exc:
>           raise ASTSafetyError(
                "cannot use --safe with this file; failed to parse source file AST: "
                f"{exc}\n"
                "This could be caused by running Black with an older Python version "
                "that does not support new syntax used in your source file."
            ) from exc
E           black.parsing.ASTSafetyError: cannot use --safe with this file; failed to parse source file AST: unexpected indent (<unknown>, line 1)
E           This could be caused by running Black with an older Python version that does not support new syntax used in your source file.

/nix/store/fdbzd7fz5p3qdvpw8a4212lcp13p9nzv-python3.11-black-24.3.0/lib/python3.11/site-packages/black/__init__.py:1528: ASTSafetyError
=============================== warnings summary ===============================
tests/test_plugin.py:8
  /build/source/tests/test_plugin.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_plugin.py::test_pylsp_format_document_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
FAILED tests/test_plugin.py::test_pylsp_format_range_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
=================== 2 failed, 19 passed, 1 warning in 0.34s ====================
/nix/store/v5lsd029lz5lfhamivbgqyp3zdv94ah2-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/sf03vzbryvdd6bc1rpc638b2znij2xir-python3.11-python-lsp-black-2.0.0.drv' failed with exit code 1;
       last 10 log lines:
       > tests/test_plugin.py:8
       >   /build/source/tests/test_plugin.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
       >     import pkg_resources
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED tests/test_plugin.py::test_pylsp_format_document_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
       > FAILED tests/test_plugin.py::test_pylsp_format_range_syntax_error - black.parsing.ASTSafetyError: cannot use --safe with this file; failed to p...
       > =================== 2 failed, 19 passed, 1 warning in 0.34s ====================
       > /nix/store/v5lsd029lz5lfhamivbgqyp3zdv94ah2-stdenv-linux/setup: line 1578: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/sf03vzbryvdd6bc1rpc638b2znij2xir-python3.11-python-lsp-black-2.0.0.drv'.
```
### Expected behavior

Should build

### Notify maintainers

@bryango

@cpcloud 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.12, NixOS, 24.05 (Uakari), 24.05.20240403.fd281bd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/n2g5cqwv8qf5p6vjxny6pg3blbdij12k-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by nazarewk**
fixes https://github.com/NixOS/nixpkgs/issues/302902

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Ramblurr — APPROVED** [review]
Result of `nixpkgs-review pr 303044` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.python-lsp-black</li>
    <li>python311Packages.python-lsp-black.dist</li>
    <li>spyder (python311Packages.spyder)</li>
    <li>spyder.dist (python311Packages.spyder.dist)</li>
  </ul>
</details>

**Comment by kirillrdy** [comment]
correct attribute is `python311Packages.python-lsp-black` , please update commit message and PR title

**Review by sumnerevans — APPROVED** [review]
Result of `nixpkgs-review pr 303044` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.python-lsp-black</li>
    <li>python311Packages.python-lsp-black.dist</li>
    <li>spyder (python311Packages.spyder)</li>
    <li>spyder.dist (python311Packages.spyder.dist)</li>
  </ul>
</details>

**Comment by rnhmjoj** [comment]
Since it has been merged, can you use https://github.com/python-lsp/python-lsp-black/commit/9298585a9d14d25920c33b188d79e820dc98d4a9 instead of the pull request number to generate the patch?
In general it's better to refer to a specific commit, so the content will not (normally) change.

**Comment by nazarewk** [comment]
I'll also drop the ` (versionOlder version "2.0.1") && (...)` to make the patches fail explicitly instead of being silently ignored

File: pkgs/development/python-modules/python-lsp-black/default.nix [hunk 1]
```diff
@@ -24,14 +24,20 @@ buildPythonPackage rec {
...
+        url = "https://github.com/python-lsp/python-lsp-black/commit/d43b41431379f9c9bb05fab158c4d97e6d515f8f.patch";
         hash = "sha256-38bYU27+xtA8Kq3appXTkNnkG5/XgrUJ2nQ5+yuSU2U=";
       })
+    ++ lib.optional
+      (with lib; versionAtLeast black.version "24.3.0")
+      (fetchpatch {
+        url = "https://github.com/python-lsp/python-lsp-black/commit/9298585a9d14d25920c33b188d79e820dc98d4a9.patch";
+        hash = "sha256-4u0VIS7eidVEiKRW2wc8lJVkJwhzJD/M+uuqmTtiZ7E=";
+      })
     ++ [ ];
```
Comment (dixslyf) [inline pkgs/development/python-modules/python-lsp-black/default.nix:41]:
```suggestion
      });
```

**Review by dixslyf — APPROVED** [review]
Result of `nixpkgs-review pr 303044` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.python-lsp-black</li>
    <li>python311Packages.python-lsp-black.dist</li>
    <li>spyder (python311Packages.spyder)</li>
    <li>spyder.dist (python311Packages.spyder.dist)</li>
  </ul>
</details>

**Pull request merged**